### PR TITLE
Refactor Harvester counter to improve performance

### LINKED ISSUE #603: [Phobos Bug] HarvesterCounter believes that the harvester in Chrono ready state is not working
### Description:
If a CMIN on its way back is walking because GAREFN is occupied, the HarvesterCounter believes that the CMIN is not working

### Phobos Build:
[Build 27](https://github.com/Phobos-developers/Phobos/releases/tag/build-27)

### Steps To Reproduce:
Make the number of CMIN more than XAREFN

### Expected Behaviour:
CMIN is still on its way back to the mine and has not changed to a non mission state, which should not be considered as not working. And the resulting high-frequency flashing of the HarvesterCounter is very annoying and meaningless (because the harvester is actually still working and does not need players to give it new instructions)

### Actual Behaviour:
The CMIN in this state is regarded as not working, and make the HarvesterCounter flashes at a high frequency

### Additional Files / Links:
HarvesterCounter_CMIN_Bug

### PR DESCRIPTION
**Pull request by chaserli**
Detached from #723 ,  restore #628 , closes #603 

* Consider non-selected guarding state as harvesting miner (can subject to change, Needs review from @Metadorius @Thrifinesma ). Returning chronominers will stop blinking the counter.

* Deprecate `TechnoTypeExt::ExtData::IsCountedAsHarvester` and merged the entire logic into `Harverster.Counted` in order to improve performance. Minor refactored related harvester-counter code.

* Add a static array of TechnoTypeClasses that are considered to be `Harvester.Counted=yes` to avoid loop through all technotypes every frame.

* Do not display harvester counter or power delta for defeated players

### CODE REVIEW

File: src/Ext/House/Body.cpp [hunk 1]
```diff
@@ -13,34 +13,23 @@ HouseExt::ExtContainer HouseExt::ExtMap;
...
-
 	int result = 0;
-	for (auto techno : *TechnoTypeClass::Array) {
-		if (auto pTechnoExt = TechnoTypeExt::ExtMap.Find(techno)) {
-			if (pTechnoExt->IsCountedAsHarvester()) {
-				result += pThis->CountOwnedAndPresent(techno);
-			}
-		}
-	}
+	for (auto pType : TechnoTypeExt::HarvesterTypes)
```
Comment (chaserli) [inline src/Ext/House/Body.cpp:31]: I got a problem here: When I tested with MO everything's fine, but when I test 
with my vanilla YR it gets an error here: either this `std::vector` or `pType` is some nonsense or `nullptr`

File: src/Ext/Techno/Body.cpp [hunk 2]
```diff
@@ -351,11 +351,19 @@ bool TechnoExt::IsHarvesting(TechnoClass* pThis)
...
+		switch (pThis->GetCurrentMission())
+		{
+		case Mission::Harvest:
+		case Mission::Unload:
+		case Mission::Enter:
+			return true;
+		case Mission::Guard: // issue#603: idk how to do better
+			if (auto pUnit = abstract_cast<UnitClass*>(pThis))
+				return !pUnit->IsSelected && pUnit->Locomotor->Is_Really_Moving_Now();
+		default:break;
```
Comment (Metadorius) [inline src/Ext/Techno/Body.cpp:365]:
```suggestion
		default: break;
```

File: src/Ext/TechnoType/Body.cpp [hunk 3]
```diff
@@ -53,18 +53,19 @@ bool TechnoTypeExt::HasSelectionGroupID(ObjectTypeClass* pType, const char* pID)
 	return (_strcmpi(id, pID) == 0);
 }
 
+[[deprecated("Use Harvester_Counted directly.")]]
```
Comment (Metadorius) [inline src/Ext/TechnoType/Body.cpp:56]: Why leave it in at all?

File: src/Ext/TechnoType/Body.cpp [hunk 4]
```diff
@@ -476,3 +486,21 @@ DEFINE_HOOK(0x679CAF, RulesClass_LoadAfterTypeData_CompleteInitialization, 0x5)
...
+{
+	GET(UnitTypeClass*, pItem, ESI);
+
+	if (auto pTypeExt = TechnoTypeExt::ExtMap.Find(pItem))
+	{
+		if (!pTypeExt->Harvester_Counted.isset() && pItem->Harvester)
+		{
+			pTypeExt->Harvester_Counted = true;
+			auto& list = RulesExt::Global()->HarvesterTypes;
+			if(!list.Contains(pItem))
```
Comment (Metadorius) [inline src/Ext/TechnoType/Body.cpp:500]:
```suggestion
			if (!list.Contains(pItem))
```

**Review by Metadorius — APPROVED** [review]

**Review by Metadorius — CHANGES_REQUESTED** [review]
Forgot to say, add yourself to authors of the features you fixed and if it was in a stable build - append the fix to the changelog.

File: src/Ext/TechnoType/Body.cpp (hunk 3 above)
Comment (chaserli) [reply src/Ext/TechnoType/Body.cpp]: Removed. Ready to go with #721

**Pull request merged**